src/hash.cpp

```diff
@@ -0,0 +1,156 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2018 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#include "mococrw/hash.h"
+#include "mococrw/util.h"
+#include "mococrw/openssl_wrap.h"
+
+namespace mococrw
+{
+using namespace openssl;
+
+static const EVP_MD* _getDigestFn (DigestTypes digestType) {
+    switch (digestType) {
+        case DigestTypes::SHA256:
+            return lib::OpenSSLLib::SSL_EVP_sha256();
+        case DigestTypes::SHA512:
+            return lib::OpenSSLLib::SSL_EVP_sha512();
+        default:
+            throw std::runtime_error("Unknown digest type");
+    }
+}
+
+static EVP_MD_CTX _initCtx (DigestTypes digestType) {
+    const EVP_MD* digestFn = _getDigestFn(digestType);
+    EVP_MD_CTX digestCtx;
+    EVP_MD_CTX_init(&digestCtx);
+    EVP_DigestInit_ex(&digestCtx, digestFn, NULL);
+    return digestCtx;
+}
+
+static void _finalizeDigest (EVP_MD_CTX& digestCtx, uint8_t* outDigest, unsigned int &outLength) {
+    EVP_DigestFinal_ex(&digestCtx, outDigest, &outLength);
+    EVP_MD_CTX_cleanup(&digestCtx);
+}
+
+static void _hash(DigestTypes digestType, const uint8_t* message, const int &length, uint8_t* outDigest, unsigned int &outLength) {
+    EVP_MD_CTX digestCtx = _initCtx(digestType);
+    EVP_DigestUpdate(&digestCtx, message, length);
+    _finalizeDigest(digestCtx, outDigest, outLength);
+}
+
+static void _hash(DigestTypes digestType, const std::string &message, std::string &outDigest) {
+    uint8_t value[Hash::lengthBytes[digestType]];
+    unsigned int length;
+    _hash(digestType, reinterpret_cast<const uint8_t*>(message.c_str()), message.length(), value, length);
+    outDigest = utility::toHex(value, length);
+}
+
+static void _hash(DigestTypes digestType, const std::vector<uint8_t> &message, std::vector<uint8_t> &outDigest) {
+    uint8_t value[Hash::lengthBytes[digestType]];
+    unsigned int length;    
+    _hash(digestType, &message[0], message.size(), value, length);
+    outDigest.insert(outDigest.begin(), value, value + length);
+    outDigest.resize(length);
```

Comment:
You end of with this strange pattern because you can't really really on the state of your output parameter. This is one of these situations where you see why this is just not preferred design.